profiling/src/php_ffi.h

```diff
@@ -0,0 +1,133 @@
+#include <SAPI.h>
+#include <Zend/zend_extensions.h>
+#include <Zend/zend_modules.h>
+#include <php.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <string.h>
+
+#include <ext/standard/info.h>
+
+/**
+ * Represents a non-owning slice of chars. The order here matches GCC std::span,
+ * which changed in v11 to be ptr-size instead of size-ptr to match iovec on
+ * most platforms.
+ *
+ * Please do not use a null pointer; instead use an empty string with a `size`
+ * of 0.
+ *
+ * todo: unify with ../components/string_view
+ */
+typedef struct datadog_php_str {
+    const char *ptr;
+    size_t size;
+} datadog_php_str;
+
+/* C11 allows a duplicate typedef provided they are the same, so this should be
+ * fine as long as we compile with C11 or higher.
+ */
+typedef ZEND_RESULT_CODE zend_result;
+
+/**
+ * Returns macro expansion of ZEND_EXTENSION_BUILD_ID, which bindgen cannot
+ * currently handle.
+ */
+const char *datadog_extension_build_id(void);
+
+/**
+ * Returns macro expansion of ZEND_MODULE_BUILD_ID, which bindgen cannot
+ * currently handle.
+ */
+const char *datadog_module_build_id(void);
+
+/**
+ * Represents a log level filter setting. Store these in uintptr_t, as the Rust
+ * type is repr(usize) aka uintptr_t.
+ * todo: unify with ../component/log
+ */
+typedef enum datadog_php_profiling_log_level {
+    DATADOG_PHP_PROFILING_LOG_LEVEL_OFF = 0,
+    DATADOG_PHP_PROFILING_LOG_LEVEL_ERROR,
+    DATADOG_PHP_PROFILING_LOG_LEVEL_WARN,
+    DATADOG_PHP_PROFILING_LOG_LEVEL_INFO,
+    DATADOG_PHP_PROFILING_LOG_LEVEL_DEBUG,
+    DATADOG_PHP_PROFILING_LOG_LEVEL_TRACE,
+} datadog_php_profiling_log_level;
+
+/**
+ * Converts the `log_level` into a `datadog_php_str`. It also guarantees NULL
+ * termination for convenience for use in debuggers, c string functions, etc.
+ */
+datadog_php_str datadog_php_profiling_log_level_to_str(uintptr_t log_level);
+
+/* Expose globals so we can bridge them with Rust. */
+// clang-format off
+ZEND_BEGIN_MODULE_GLOBALS(datadog_php_profiling)
+    bool profiling_enabled;
+    bool profiling_experimental_cpu_time_enabled;
+    uint32_t interrupt_count;
+
+    // datadog_php_profiling_log_level maps to LevelFilter which is repr(usize)
+    uintptr_t profiling_log_level;
+
+    zend_bool *vm_interrupt_addr;
+
+    // The strings will be interned but potentially only for the request, so be
+    // careful to not use them outside of a request (such as from other
+    // threads).
+    datadog_php_str env;
+    datadog_php_str service;
+    datadog_php_str version;
+    datadog_php_str trace_agent_host;
+    datadog_php_str trace_agent_port;
+    datadog_php_str trace_agent_url;
```

Comment:
Can these change from request-to-request (particularly wondering about sapi_getenv() ... the libc getenv() shouldn't really change during execution).

If not, then perhaps we should hold a void* to a `Box::into_raw()` of a Rust structure so we don't have to bounce these strings through FFI so much just for them to end back up in a `Vec<Tag>`

Reply:
They can change from request to request, yes. For instance, in Apache you can configure env vars to have different values for different directories, think `DD_SERVICE=myservice` but in `/admin` it's `DD_SERVICE=myservice-admin` or something.

The code is partly set up the way it is because PHP's request globals have support for being tied to `.ini` configuration. I had previously tried to add `.ini` support, but hit an issue with certain configs like `datadog.service` being shared by the tracer and profiler, and the engine does not like this (it's not a C vs Rust issue, but a PHP extension issue). So, I pulled out all of the _directly_ dead code but some of the influence is still there. I think I need to support `.ini` configuration so I'm not sure it's worth making the Rust side of things cleaner at the moment since I'm highly likely to undo that work for `.ini`. Dunno, I'll think about it.